File: backend/app/schemas/suggestion.py
"""Suggestion-related DTOs."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

SuggestionDisplayStatusCode = Literal["pending", "exported", "archived", "error"]


class AllocationExplanationOut(BaseModel):
    allocation_mode: str
    matched_order_qty: int
    unknown_order_qty: int
    eligible_warehouses: list[str]


class SuggestionOut(BaseModel):
    id: int
    status: str
    triggered_by: str
    total_items: int
    procurement_item_count: int = 0
    restock_item_count: int = 0
    procurement_snapshot_count: int = 0
    restock_snapshot_count: int = 0
    archived_trigger: str | None = None
    # 派生状态，供历史页直接展示（未导出 / 已导出 / 已归档）
    procurement_display_status: str = "未导出"
    restock_display_status: str = "未导出"
    # 机器可读的状态码，供前端按 code 映射 tag 色 / i18n（label 保留 UX 兼容）
    procurement_display_status_code: SuggestionDisplayStatusCode = "pending"
    restock_display_status_code: SuggestionDisplayStatusCode = "pending"
    global_config_snapshot: dict[str, Any]
    created_at: datetime
    archived_at: datetime | None = None

    model_config = {"from_attributes": True}


class SuggestionItemOut(BaseModel):
    id: int
    commodity_sku: str
    commodity_name: str | None = None
    main_image: str | None = None
    total_qty: int
    country_breakdown: dict[str, Any]
    warehouse_breakdown: dict[str, Any]
    restock_dates: dict[str, str | None] = Field(default_factory=dict)
    allocation_snapshot: dict[str, AllocationExplanationOut] | None = None
    velocity_snapshot: dict[str, Any] | None = None
    sale_days_snapshot: dict[str, Any] | None = None
    urgent: bool
    purchase_qty: int
    purchase_date: date | None = None
    procurement_export_status: str
    procurement_exported_snapshot_id: int | None = None
    procurement_exported_at: datetime | None = None
    restock_export_status: str
    restock_exported_snapshot_id: int | None = None
    restock_exported_at: datetime | None = None

    model_config = {"from_attributes": True}


class SuggestionDetailOut(SuggestionOut):
    items: list[SuggestionItemOut]


class SuggestionListOut(BaseModel):
    items: list[SuggestionOut]
    total: int
    page: int
    page_size: int


class SuggestionItemPatch(BaseModel):
    total_qty: int | None = Field(default=None, ge=0)
    purchase_qty: int | None = Field(default=None, ge=0)
    purchase_date: date | None = None
    country_breakdown: dict[str, int] | None = None
    warehouse_breakdown: dict[str, dict[str, int]] | None = None

    @model_validator(mode="after")
    def _values_non_negative(self) -> SuggestionItemPatch:
        if self.country_breakdown:
            for country, qty in self.country_breakdown.items():
                if qty < 0:
                    raise ValueError(f"country_breakdown[{country}] 不可为负")
        if self.warehouse_breakdown:
            for country, warehouse_map in self.warehouse_breakdown.items():
                for warehouse_id, qty in warehouse_map.items():
                    if qty < 0:
                        raise ValueError(f"warehouse_breakdown[{country}][{warehouse_id}] 不可为负")
        return self

File: frontend/src/api/suggestion.ts
import type { SortOrder } from '@/utils/tableSort'

import client from './client'
import type { PageResult } from './data'

export interface AllocationExplanation {
  allocation_mode: 'matched' | 'fallback_even' | 'no_warehouse'
  matched_order_qty: number
  unknown_order_qty: number
  eligible_warehouses: string[]
}

export interface CalculationInputsSnapshot {
  version: number
  generated_at: string
  demand_date: string
  target_days: number
  demand_days: number
  effective_target_days: number
  safety_stock_days: number
  purchase: {
    country_restock_qty_total: number
    country_restock_qty_by_country: Record<string, number>
    daily_velocity_total: number
    daily_velocity_by_country: Record<string, number>
    safety_stock_qty: number
    local_stock_available: number
    local_stock_reserved: number
    local_stock_total: number
    raw_purchase_qty: number
    final_purchase_qty: number
  }
  restock: {
    countries: Record<
      string,
      {
        effective_target_days: number
        daily_velocity: number
        overseas_available: number
        overseas_reserved: number
        in_transit: number
        overseas_stock_total: number
        target_stock_qty: number
        raw_restock_qty: number
        final_restock_qty: number
        sale_days: number | null
        restock_date: string | null
      }
    >
  }
}

export type SuggestionDisplayStatus = '未导出' | '已导出' | '已归档'
// 机器可读 code（4 档），供 tag 色映射 / i18n；label 保留 3 档 UX 兼容
export type SuggestionDisplayStatusCode = 'pending' | 'exported' | 'archived' | 'error'

export interface Suggestion {
  id: number
  status: 'draft' | 'archived' | 'error'
  triggered_by: string
  total_items: number
  procurement_item_count: number
  restock_item_count: number
  procurement_snapshot_count: number
  restock_snapshot_count: number
  archived_trigger: string | null
  procurement_display_status: SuggestionDisplayStatus
  restock_display_status: SuggestionDisplayStatus
  procurement_display_status_code: SuggestionDisplayStatusCode
  restock_display_status_code: SuggestionDisplayStatusCode
  global_config_snapshot: Record<string, unknown>
  created_at: string
  archived_at: string | null
}

export interface SuggestionItem {
  id: number
  commodity_sku: string
  commodity_name: string | null
  main_image: string | null
  total_qty: number
  country_breakdown: Record<string, number>
  warehouse_breakdown: Record<string, Record<string, number>>
  restock_dates: Record<string, string | null>
  allocation_snapshot: Record<string, AllocationExplanation> | null
  velocity_snapshot: Record<string, number> | null
  sale_days_snapshot: Record<string, number> | null
  calculation_inputs_snapshot: CalculationInputsSnapshot | null
  calculation_warnings: Array<{
    code: string
    country: string
    reason: string | null
    message: string
  }>
  urgent: boolean
  purchase_qty: number
  procurement_export_status: 'pending' | 'exported'
  procurement_exported_snapshot_id: number | null
  procurement_exported_at: string | null
  restock_export_status: 'pending' | 'exported'
  restock_exported_snapshot_id: number | null
  restock_exported_at: string | null
}

export interface SuggestionDetail extends Suggestion {
  items: SuggestionItem[]
}

export interface SuggestionItemPatch {
  total_qty?: number
  purchase_qty?: number
  country_breakdown?: Record<string, number>
  warehouse_breakdown?: Record<string, Record<string, number>>
}

export async function listSuggestions(params: {
  status?: string
  display_status?: 'pending' | 'exported' | 'archived' | 'error'
  date_from?: string
  date_to?: string
  sku?: string
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: SortOrder
}): Promise<PageResult<Suggestion>> {
  const { data } = await client.get('/api/suggestions', { params })
  return data
}

export async function getCurrentSuggestion(): Promise<SuggestionDetail> {
  const { data } = await client.get<SuggestionDetail>('/api/suggestions/current')
  return data
}

export async function getSuggestion(id: number): Promise<SuggestionDetail> {
  const { data } = await client.get<SuggestionDetail>(`/api/suggestions/${id}`)
  return data
}

export async function patchSuggestionItem(
  suggestionId: number,
  itemId: number,
  patch: SuggestionItemPatch,
): Promise<SuggestionItem> {
  const { data } = await client.patch<SuggestionItem>(
    `/api/suggestions/${suggestionId}/items/${itemId}`,
    patch,
  )
  return data
}

export async function recalculateSuggestionItem(
  suggestionId: number,
  itemId: number,
): Promise<SuggestionItem> {
  const { data } = await client.post<SuggestionItem>(
    `/api/suggestions/${suggestionId}/items/${itemId}/recalculate`,
  )
  return data
}

export async function deleteSuggestion(suggestionId: number): Promise<void> {
  await client.delete(`/api/suggestions/${suggestionId}`)
}
